### PR TITLE
Blacklist network interfaces from the net admin

### PR DESCRIPTION
**Pull request by ibinshtok**
Signed-off-by: ibinshtok <ilya.binshtok@eurotech.com>

### CODE REVIEW

File: kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/system/SystemService.java
Comment (MMaiero) [inline kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/system/SystemService.java]: I believe a version needs to be set here

File: kura/org.eclipse.kura.core/OSGI-INF/system.xml
Comment (MMaiero) [inline kura/org.eclipse.kura.core/OSGI-INF/system.xml]: If it is not needed anymore, let's remove it

**Comment by MMaiero** [comment]
Please have also a look why the tests are failing. Thanks

File: kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/system/SystemService.java
Comment (ibinshtok) [reply kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/system/SystemService.java]: done

File: kura/org.eclipse.kura.core/OSGI-INF/system.xml
Comment (ibinshtok) [reply kura/org.eclipse.kura.core/OSGI-INF/system.xml]: done

**Comment by ibinshtok** [comment]
I committed my Lint fixes and tested on RG 10-20 and 20-25. It worked fine but my changes created a few test errors in org.eclipse.kura.core.ssl.test and org.eclipse.kura.core.system.test. @markoer, could you take a look at this tomorrow if you have a chance, please?

File: kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/system/SystemService.java
Comment (MMaiero) [inline kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/system/SystemService.java]: This is usually related to the API package version. So, it should be 1.4.0

**Review by markoer — COMMENTED** [review]
Which PR/functionality would cover Kura not configuring eth0 as DHCP during installation? If this one, then installation script would have to check this new property kura.net.admin.interface.blacklist and update the existing /etc/network/interfaces instead of overwriting it. Assuming, of course, that the installation package would already contain the updated kura.properties...
So, where does this PR end and some other begin?

**Review by markoer — APPROVED** [review]

**Comment by MMaiero** [comment]
@markoer tests are still failing. Are we planning on having changes in them before merging?
Thanks

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/ssl/SslManagerServiceImpl.java
Comment (markoer) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/ssl/SslManagerServiceImpl.java]: It seems not to be a good idea to open the same file for reading and writing at the same time. If you move the output stream back down, it should work. You can add another try-with-resources, there.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/ssl/SslManagerServiceImpl.java
Comment (markoer) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/ssl/SslManagerServiceImpl.java]: See the comment above.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/ssl/SslManagerServiceImpl.java
Comment (markoer) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/ssl/SslManagerServiceImpl.java]: See the comment above.

**Review by markoer — CHANGES_REQUESTED** [review]
Once these and Matteo's proposed changes are implemented, I approve.

**Comment by markoer** [comment]
I see reviews are not really visible per-commit... So, I approve the changes after the code is fixed. The tests are still OK with the proposed code updates.
And the `@since` javadoc still needs to be updated, too.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/ssl/SslManagerServiceImpl.java
Comment (ibinshtok) [reply kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/ssl/SslManagerServiceImpl.java]: Yes, I've already rolled this back on Friday w/o committing since I did not understand what exactly happens. Thanks!

**Comment by ibinshtok** [comment]
@cdealti please take a loot at this PR. At this point, I believe, the only issue is in the org.eclipse.kura.core.test.AllCoreTests as shown below. It looks like the setConfigService() method is not being invoked and as a result the s_configService is null in the setUpClass() method. But I don't know why.

[INFO] org.eclipse.kura.core.system.test .................. SUCCESS [  7.036 s]
[INFO] org.eclipse.kura.core.test ......................... FAILURE [ 20.662 s]

ibinshtok@T560:~/eurotechdev/kura-ilya/kura/kura/test/org.eclipse.kura.core.test/target/surefire-reports$ more org.eclipse.kura.core.test.AllCoreTests.txt
-------------------------------------------------------------------------------
Test set: org.eclipse.kura.core.test.AllCoreTests
-------------------------------------------------------------------------------
Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 11.223 sec <<< FAILURE! - in org.eclipse.kura.core.test.AllCoreTests
org.eclipse.kura.core.test.AllCoreTests  Time elapsed: 11.198 sec  <<< ERROR!
java.lang.Exception: Failed to reconfigure the broker settings - failing out
	at org.eclipse.kura.core.test.AllCoreTests.setUpClass(AllCoreTests.java:94)

org.eclipse.kura.core.test.AllCoreTests  Time elapsed: 11.222 sec  <<< ERROR!
java.lang.NullPointerException: null
	at org.eclipse.kura.core.test.AllCoreTests.tearDownClass(AllCoreTests.java:149)

**Comment by cdealti** [comment]
@ibinshtok before starting or restarting the review, I'd like to restate the requirements.

1. If and interface is blacklisted, it will be filtered out by Kura at the lowest possible layer. In principle I would not expect changes to the Web UI or other parts of Kura. The interface is simply not available to the net admin or to applications.

2. Kura must preserve the configuration for the interfaces it does not manage. For example, if an interface usb0 is blacklisted, Kura must preserve its configuration in /etc/network/interfaces or other configuration files.

For the time being I would only focus on that.

**Review by markoer — APPROVED** [review]

**Comment by markoer** [comment]
And Cristiano's requirements still stand, despite the approval of changes ;)

**Comment by ibinshtok** [comment]
@cdealti The changes to the Web UI I made are not functional changes just an attempt to fig Lint issues. I touched Web UI files because of Lint issues addressed in the SystemServiceImpl. 
Functional changes are in the SystemServiceImpl and NetworkServiceImpl in the getAllNetworkInterfaceNames() method.

**Comment by cdealti** [comment]
@ibinshtok I really appreciate the effort of writing the Python script for preserving the network configurations during installation. However I wonder if we could keep it on hold for a while.

I'd really like if we could avoid installing system configuration with the installer files and having Kura generating the them at the first boot.

The general idea is that Kura should be able to configure everything from the configuration snapshot.

So for the time being, I'd revert the Python addition.

**Comment by cdealti** [comment]
@ibinshtok @markoer  this seems to be no longer needed.
Can we close it and reopen if needed?

**Comment by markoer** [comment]
The only thing missing in the unmanaged interfaces is the script that merges the `/etc/network/interfaces` files during Kura installation. With it installation doesn't overwrite the interfaces, that are not expected on the system, e.g. additional usb2eth interfaces.

**Comment by cdealti** [comment]
> The only thing missing in the unmanaged interfaces is the script that merges the /etc/network/interfaces files during Kura installation.

This requirement is quite unique. Usually the installer just replaces/overwrites existing configuration files. I'm not sure whether we should make an exception to this rule.
Maybe, if this is needed, this is the job of an external script

**Pull request closed without merging**
